src/storage/gc_worker.rs

```diff
@@ -0,0 +1,306 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::engine::{Engine, Error as EngineError, ScanMode, StatisticsSummary};
+use super::metrics::*;
+use super::mvcc::{MvccReader, MvccTxn};
+use super::txn::{GC_BATCH_SIZE, GC_LOG_DELETED_VERSION_THRESHOLD, GC_LOG_FOUND_VERSION_THRESHOLD};
+use super::{Callback, Error, Key, Result};
+use kvproto::kvrpcpb::Context;
+use std::fmt::{self, Display, Formatter};
+use std::mem;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+use util::time::{duration_to_sec, SlowTimer};
+use util::worker::{self, Builder, Runnable, ScheduleError, Worker};
+
+pub const GC_MAX_PENDING_TASKS: usize = 2;
+const GC_SNAPSHOT_TIMEOUT_SECS: u64 = 10;
+const GC_TASK_SLOW_SECONDS: u64 = 30;
+
+struct GCTask {
+    pub ctx: Context,
+    pub safe_point: u64,
+    pub callback: Callback<()>,
+}
+
+impl Display for GCTask {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        let epoch = format!("{:?}", self.ctx.region_epoch.as_ref());
+        f.debug_struct("GCTask")
+            .field("region", &self.ctx.get_region_id())
+            .field("epoch", &epoch)
+            .field("safe_point", &self.safe_point)
+            .finish()
+    }
+}
+
+/// `GCRunner` is used to perform GC on the engine
+struct GCRunner<E: Engine> {
+    engine: E,
+    ratio_threshold: f64,
+
+    stats: StatisticsSummary,
+}
+
+impl<E: Engine> GCRunner<E> {
+    pub fn new(engine: E, ratio_threshold: f64) -> GCRunner<E> {
+        GCRunner {
+            engine,
+            ratio_threshold,
+
+            stats: StatisticsSummary::default(),
+        }
+    }
+
+    fn get_snapshot(&self, ctx: &mut Context) -> Result<E::Snap> {
+        let timeout = Duration::from_secs(GC_SNAPSHOT_TIMEOUT_SECS);
```

Comment:
This seems to be duplicated with `engine.snapshot`, only differs in the amount of timeout. Consider extract to a function?

Reply:
I want to get the `cb_ctx`, which is dropped by `engine.snapshot`.
It may be ok to let `engine.snapshot` return it, in other words, return `Result<(Self::Snap, CbContext)>` from `engine.snapshot`. If you agree, I'll change it later.

Reply:
Ok, seems that it would be better to keep the code as current.